### PR TITLE
Serdes v2 import export cache

### PR DESCRIPTION
**Pull request by john-metabase**
This adds metadata to functions which can be safely cached during import and export, and adds a macro to temporarily re-bind those functions with memoized versions while executing some code.

- - -
This change is [Reviewable](https://reviewable.io/reviews/metabase/metabase/29506)

### CODE REVIEW

**Comment by john-metabase** [comment]
This is a super naive approach, and I hate adding earmuffs to all of those functions just to make it correct, in this case. I originally started down the path of passing some kind of a mutable cache value or map of functions but that was way uglier to make it work across all of the different places we could optimize.

**Review by bshepherdson — CHANGES_REQUESTED** [review]
> Reviewed 1 of 5 files at r1, 20 of 20 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/metabase/metabase/29506)* status: all files reviewed, 1 unresolved discussion (waiting on @camsaul, @john-metabase, and @luizarakaki)

___
*[`src/metabase/models/serialization.clj` line 984 at r2](https://reviewable.io/reviews/metabase/metabase/29506#-NRZoQ3IDkY9qt6pN8bi:-NRZoQ3IDkY9qt6pN8bj:b-w6ehts) ([raw file](https://github.com/metabase/metabase/blob/91136e7ea7159aa01f3bc277feb238059215d53d/src/metabase/models/serialization.clj#L984)):*
> ```Clojure
>          (filter some?)
>          (reduce set/union))))
> ```

I think there's a tidier way of doing this. We already import core.memoize, so let's use that (and its tangible memoization caches) to implement `with-cache` without having to make all the vars `^:dynamic`. Something like this (untested)

```clojure
(def import-field-fk
  "Given a `field_id` as exported by [[export-field-fk]], resolve it back into a numeric `field_id`."
  (memoize/memo (fn [field] #_import-field-fk-body-here)))  

(defmaco with-cache [& body]
  `(do
     (doseq [[var-sym var] (ns-interns ns*) :when (memoize/memoized? var)]
        (memoize/memo-clear! var))
     ~@body))
```



_Code quote:_
```Clojure
(defmacro with-cache
  "Runs body with all functions marked with ::cache re-bound to memoized versions for performance."
  [& body]
  (let [ns* 'metabase.models.serialization]
    `(binding ~(reduce into []
                       (for [[var-sym var] (ns-interns ns*)
                             :when (::cache (meta var))
                             :let  [fq-sym (symbol (name ns*) (name var-sym))]]
                         [fq-sym `(memoize ~fq-sym)]))
       ~@body)))
```

**Review by bshepherdson — COMMENTED** [review]
See my comment for a possible other approach. I agree that the earmuffs sucks, and that there's got to be a better way.

I wish I could supply my own Atoms to `core.memoize` but it doesn't seem like it really supports that. But clearing them on entry to `with-cache` works well enough.

> *[Reviewable](https://reviewable.io/reviews/metabase/metabase/29506#-:-NR_4VtZ1Wn2xKm1gVn6:b-gsajzh)* status: all files reviewed, 1 unresolved discussion (waiting on @camsaul, @john-metabase, and @luizarakaki)

**Comment by john-metabase** [comment]
Thanks @bshepherdson, I think I found a middle ground, still using metadata to mark which functions should be memoized, so we can keep the nice `defn` style.

File: src/metabase/models/serialization.clj [hunk 1]
```diff
@@ -981,3 +982,21 @@
     (->> (concat vis-column-settings [(mbql-deps viz) link-card-deps])
          (filter some?)
          (reduce set/union))))
+
+(def ^:private cached-vars
+  (filter (comp ::cache meta) (vals (ns-interns *ns*))))
+
+(defn init-cache
```
Comment (john-metabase) [inline src/metabase/models/serialization.clj:989]: PS, this is public because the macro needs to invoke this function where it is expanded.

**Review by bshepherdson — APPROVED** [review]
> Reviewed 17 of 18 files at r3, 1 of 1 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/metabase/metabase/29506)* status: :shipit: complete! all files reviewed, all discussions resolved (waiting on @camsaul and @luizarakaki)

*[a discussion](https://reviewable.io/reviews/metabase/metabase/29506#gh-1150833633:-NRdmhoq5nqKAWEr6F8m:b-dxcbts) (no related file):*
<details><summary><i>Previously, john-metabase wrote…</i></summary><blockquote>

*(Reviewable was [unable](https://github.com/Reviewable/Reviewable/issues/14) to map this GitHub inline comment thread to the right spot — sorry!)*

PS, this is public because the macro needs to invoke this function where it is expanded.
</blockquote></details>

Makes sense.

___
___
*[`src/metabase/models/serialization.clj` line 995 at r4](https://reviewable.io/reviews/metabase/metabase/29506#-NRdmMBF138rEavHCuQ5:-NRdmMBF138rEavHCuQ6:b-22jucp) ([raw file](https://github.com/metabase/metabase/blob/b85bdc459c6ed33cbc01172a65282fb8268dc5e7/src/metabase/models/serialization.clj#L995)):*
> ```Clojure
>     (if (memoize/memoized? @v)
>       (memoize/memo-clear! @v)
>       (alter-var-root v memoize/memo))))
> ```

Sweet, I like this approach.



_Code quote:_
```Clojure
  (doseq [v cached-vars]
    (if (memoize/memoized? @v)
      (memoize/memo-clear! @v)
      (alter-var-root v memoize/memo))))
```

**Review by camsaul — APPROVED** [review]

**Review by bshepherdson — APPROVED** [review]
> Reviewed 19 of 19 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/metabase/metabase/29506)* status: :shipit: complete! all files reviewed, all discussions resolved (waiting on @luizarakaki)

**Pull request merged**